README.md
![Magic](./magic.gif)

# orthomagic4ps
Magic Photoshop scripts to make editing orthophotos easier. Intended as a supplement to [Ortho4XP](https://github.com/oscarpilote/Ortho4XP), as such knowing your way around it is required.

## Install
Download the latest version from the [releases](https://github.com/MorisatoK/orthomagic4ps/releases) page. Save it to a place of your liking and where you can access it from Photoshop.

## Usage
See the [wiki](https://github.com/MorisatoK/orthomagic4ps/wiki) for detailed instructions and recommendations.

## Changes
See the [changelog](./CHANGELOG.md)

CHANGELOG.md
# CHANGELOG

## 1.0.0 (2020-06-07)

Feature:
  - Export Magic

Behind the scene changes for code sanity.

## 0.2.0 (2020-06-04)

Feature:
  - Dialog to add guides to the document after placing the textures

Fix:
  - fix offsets when placing higher zoom level textures than documents base zoom level
  - additional sanity check when parsing info from document name
  - add missing types

## 0.1.0 (2020-06-01)

Initial Release


.eslintrc.js
// eslint-disable-next-line no-undef
module.exports = {
    root: true,
    parser: '@typescript-eslint/parser',
    parserOptions: {
        // eslint-disable-next-line no-undef,  @typescript-eslint/no-unsafe-assignment
        tsconfigRootDir: __dirname,
        project: ['./tsconfig.json'],
    },
    plugins: ['@typescript-eslint'],
    extends: [
        'eslint:recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'prettier/@typescript-eslint',
    ],
    rules: {
        '@typescript-eslint/array-type': ['error', {default: 'array-simple'}],
        '@typescript-eslint/consistent-type-assertions': [
            'error',
            {assertionStyle: 'angle-bracket', objectLiteralTypeAssertions: 'allow-as-parameter'},
        ],
        '@typescript-eslint/consistent-type-definitions': ['error', 'interface'],
        '@typescript-eslint/explicit-function-return-type': ['error'],
        '@typescript-eslint/explicit-member-accessibility': [
            'error',
            {
                accessibility: 'explicit',
                overrides: {
                    constructors: 'no-public',
                },
            },
        ],
        '@typescript-eslint/method-signature-style': ['error'],
        camelcase: 'off',
        '@typescript-eslint/naming-convention': [
            'error',
            {
                selector: 'typeParameter',
                format: ['PascalCase'],
                prefix: ['T'],
            },
            {
                selector: 'interface',
                format: ['PascalCase'],
                custom: {
                    regex: '^I[A-Z].+',
                    match: true,
                },
            },
        ],
        '@typescript-eslint/no-base-to-string': ['error'],
        '@typescript-eslint/prefer-for-of': ['error'],
        '@typescript-eslint/type-annotation-spacing': ['error'],
    },
};


src/importmagic.ts
class ImportMagic {
    private readonly TEXTURE_SIZE = 4096;
    private originalRulerUnits: Units | undefined;
    private documentName: string | undefined;
    private tileInfo: ITileInfo | undefined;
    private movedLayers = 0;
    private skippedLayers = 0;

    constructor() {
        if (!app.documents.length) this.fail('No document to work with.');

        if (this.hasBackgroundLayer()) this.fail('Documents with a background layer are not supported.');

        this.initRulerUnits();
        this.getDocumentName();
        this.getTileInfo();
        this.setDocDimensions();
        this.moveTextures();
        this.drawGuidesConfirmation();
        this.finish();
    }

    private hasBackgroundLayer(): boolean {
        // Ugh, Adobe, what are you doing?!
        try {
            app.activeDocument.backgroundLayer;

            return true;
        } catch (error) {
            return false;
        }
    }

    private initRulerUnits(): void {
        this.originalRulerUnits = app.preferences.rulerUnits;
        app.preferences.rulerUnits = Units.PIXELS;
    }

    private restoreRulerUnits(): void {
        if (typeof this.originalRulerUnits !== 'undefined') app.preferences.rulerUnits = this.originalRulerUnits;
    }

    private getDocumentName(): void {
        this.documentName = this.stripExtensionFromName(app.activeDocument.name);
    }

    private stripExtensionFromName(name: string): string {
        return name.split('.')[0];
    }

    private getTileInfo(): void {
        try {
            if (typeof this.documentName === 'undefined') throw '';

            this.tileInfo = this.parseTileInfoFromDocName(this.documentName);
        } catch (error) {
            this.fail('Could not parse needed info from filename. Make sure it is named like "+33+126_17".');
        }
    }

    private parseTileInfoFromDocName(docName: string): ITileInfo {
        const regex = /([+-]\d{1,3})/g;
        const latLon: RegExpMatchArray | null = docName.match(regex);
        const zoom: string | undefined = docName.split('_')[1];

        if (latLon === null || latLon?.length <= 1 || typeof zoom === 'undefined') throw '';

        return {
            lat: parseInt(latLon[0], 10),
            lon: parseInt(latLon[1], 10),
            zoom: parseInt(zoom, 10),
        };
    }

    private setDocDimensions(): void {
        const texturesCounts: ITexCount = this.getTexturesCounts();
        const xDimension: number = this.calcDimension(texturesCounts.x);
        const yDimension: number = this.calcDimension(texturesCounts.y);

        app.activeDocument.resizeCanvas(xDimension, yDimension);
    }

    private getTexturesCounts(): ITexCount {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const textureCoords: ITexCoords = this.getTextureCoords(this.tileInfo!.zoom);
        const xTexturesCount: number = (textureCoords.right - textureCoords.left) / 16 + 1;
        const yTexturesCount: number = (textureCoords.bottom - textureCoords.top) / 16 + 1;

        return {x: xTexturesCount, y: yTexturesCount};
    }

    private calcDimension(texturesCount: number): number {
        return texturesCount * this.TEXTURE_SIZE;
    }

    private getTextureCoords(zoom: number): ITexCoords {
        /* eslint-disable @typescript-eslint/no-non-null-assertion */
        const leftTop: ITexCoord = this.wgs84_to_texture(this.tileInfo!.lat + 1, this.tileInfo!.lon, zoom);
        const rightBottom: ITexCoord = this.wgs84_to_texture(this.tileInfo!.lat, this.tileInfo!.lon + 1, zoom);
        /* eslint-enable @typescript-eslint/no-non-null-assertion */

        return {
            left: leftTop.x,
            top: leftTop.y,
            right: rightBottom.x,
            bottom: rightBottom.y,
        };
    }

    private wgs84_to_texture(lat: number, lon: number, zoom: number): ITexCoord {
        const zoomDivider: number = this.getPositionDivider(zoom);
        const ratio_x: number = lon / 180;
        const ratio_y: number = Math.log(Math.tan(((90 + lat) * Math.PI) / 360)) / Math.PI;
        const pic_x: number = (ratio_x + 1) * Math.pow(2, zoom + 7);
        const pic_y: number = (1 - ratio_y) * Math.pow(2, zoom + 7);
        const tile_x: number = Math.floor(pic_x / 256);
        const tile_y: number = Math.floor(pic_y / 256);
        const tex_x: number = Math.floor(tile_x / zoomDivider) * zoomDivider;
        const tex_y: number = Math.floor(tile_y / zoomDivider) * zoomDivider;

        return {x: tex_x, y: tex_y};
    }

    private moveTextures(): void {
        const layers: Layers = app.activeDocument.layers;

        // Type 'Layers' is not an array type or a string type or does not have a '[Symbol.iterator]()' method that returns an iterator.
        // eslint-disable-next-line @typescript-eslint/prefer-for-of
        for (let i = 0; i < layers.length; i++) {
            const texInfo: ITexInfo | null = this.getLayerInfo(layers[i]);

            if (texInfo === null) {
                this.skippedLayers++;
                continue;
            }

            const textureCoords: ITexCoords = this.getTextureCoords(texInfo.tex_zoom);
            const xPos: number = (texInfo.tex_x - textureCoords.left) / this.getPositionDivider(texInfo.tex_zoom);
            const yPos: number = (texInfo.tex_y - textureCoords.top) / this.getPositionDivider(texInfo.tex_zoom);

            this.moveLayerTo(layers[i], xPos, yPos);

            // For placing higher ZL than document
            if (texInfo.tex_zoom !== this.tileInfo?.zoom) {
                this.scaleLayer(layers[i], this.getPositionDividerMultiplier(texInfo.tex_zoom));
            }
        }
    }

    private getPositionDivider(zoom: number): number {
        const positionDividerBaseValue = 16;

        return positionDividerBaseValue * this.getPositionDividerMultiplier(zoom);
    }

    private getPositionDividerMultiplier(zoom: number): number {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return Math.pow(2, zoom - this.tileInfo!.zoom);
    }

    private getLayerInfo(layer: Layer): ITexInfo | null {
        const splitName: string[] = layer.name.split('_');

        // Exactly three parts are expected
        if (splitName.length !== 3) return null;

        const tex_y: number = parseInt(splitName[0], 10);
        const tex_x: number = parseInt(splitName[1], 10);
        const tex_zoom: number = parseInt(splitName[2].slice(-2), 10);

        // All three parts are expected to be numbers
        if (isNaN(tex_x) || isNaN(tex_y) || isNaN(tex_zoom)) return null;

        // Some sanity check for supported zoom levels
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        if (tex_zoom < this.tileInfo!.zoom || tex_zoom > 19) return null;

        return {tex_x, tex_y, tex_zoom};
    }

    private moveLayerTo(layer: Layer, x: number, y: number): void {
        const bounds: UnitRect = layer.bounds;
        const x_pos: number | UnitValue = x * this.TEXTURE_SIZE - <number>bounds[0];
        const y_pos: number | UnitValue = y * this.TEXTURE_SIZE - <number>bounds[1];

        // Only move layer it it needs to move
        if (x_pos === 0 && y_pos === 0) {
            this.skippedLayers++;
            return;
        }

        layer.translate(-x_pos, -y_pos);
        this.movedLayers++;
    }

    private scaleLayer(layer: Layer, divider: number): void {
        const bounds: UnitRect = layer.bounds;
        const width: number = <number>bounds[2] - <number>bounds[0];

        // Only resize if it is not yet resized
        if (width !== this.TEXTURE_SIZE) return;

        const imageScalePercentage: number = (1 / divider) * 100;
        layer.resize(imageScalePercentage, imageScalePercentage, AnchorPosition.TOPLEFT);
    }

    private drawGuidesConfirmation(): void {
        const confirmation: boolean = confirm('Draw guides?', true, 'Magic Guides');

        if (confirmation) this.drawGuides();
    }

    private drawGuides(): void {
        const texturesCounts: ITexCount = this.getTexturesCounts();
        const guides: IExtGuides = <IExtGuides>app.activeDocument.guides;

        this.deleteGuides();

        for (let i = 1; i < texturesCounts.x; i++) guides.add(Direction.VERTICAL, i * this.TEXTURE_SIZE);

        for (let i = 1; i < texturesCounts.y; i++) guides.add(Direction.HORIZONTAL, i * this.TEXTURE_SIZE);
    }

    private deleteGuides(): void {
        const guides: IExtGuides = <IExtGuides>app.activeDocument.guides;

        while (guides.length > 0) {
            // Type 'IExtGuides' is not an array type or a string type or does not have a '[Symbol.iterator]()' method that returns an iterator.
            // eslint-disable-next-line @typescript-eslint/prefer-for-of
            for (let i = 0; i < guides.length; i++) {
                const guide: Guide = guides[i];
                guide.remove();
            }
        }
    }

    private fail(error: string): void {
        this.restoreRulerUnits();

        throw new Error(error);
    }

    private finish(): void {
        this.restoreRulerUnits();

        alert(`Operation finished.\n\nMoved Layers: ${this.movedLayers}\nSkipped Layers: ${this.skippedLayers}`);
    }
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const importMagic: ImportMagic = new ImportMagic();


src/exportmagic.ts
class ExportMagic {
    private readonly MAX_ZOOM: number = 19;
    private readonly PROC_START_SNAPSHOT_NAME: string = 'PROCSTART';
    private readonly EXPORT_OPTIONS: JPEGSaveOptions = new JPEGSaveOptions();
    private cTID: (str: string) => number = (str: string) => app.charIDToTypeID(str);
    private sTID: (str: string) => number = (str: string) => app.stringIDToTypeID(str);

    private originalRulerUnits: Units | undefined;
    private documentName: string | undefined;
    private tileInfo: ITileInfo | undefined;
    private zoomGroups: ITexZoomGroups = {};

    constructor() {
        this.checkPreconditions();
        this.initRulerUnits();
        this.setExportOptions();
        this.getDocumentName();
        this.getTileInfo();
        this.groupTexByZoom();
        this.processTextures();
        this.finish();
    }

    private checkPreconditions(): void {
        if (!app.documents.length) this.fail('No document to work with.');

        if (!app.activeDocument.saved) this.fail('Please save your document first before starting the export.');

        if (this.hasBackgroundLayer()) this.fail('Documents with a background layer are not supported.');
    }

    private hasBackgroundLayer(): boolean {
        // Ugh, Adobe, what are you doing?!
        try {
            app.activeDocument.backgroundLayer;

            return true;
        } catch (error) {
            return false;
        }
    }

    private initRulerUnits(): void {
        this.originalRulerUnits = app.preferences.rulerUnits;
        app.preferences.rulerUnits = Units.PIXELS;
    }

    private restoreRulerUnits(): void {
        if (typeof this.originalRulerUnits !== 'undefined') app.preferences.rulerUnits = this.originalRulerUnits;
    }

    private setExportOptions(): void {
        this.EXPORT_OPTIONS.embedColorProfile = true;
        // Somewhat higher than input JPEG, because lossyness
        this.EXPORT_OPTIONS.quality = 9;
        this.EXPORT_OPTIONS.formatOptions = FormatOptions.STANDARDBASELINE;
        this.EXPORT_OPTIONS.matte = MatteType.NONE;
    }

    private getDocumentName(): void {
        this.documentName = this.stripExtensionFromName(app.activeDocument.name);
    }

    private stripExtensionFromName(name: string): string {
        return name.split('.')[0];
    }

    private getTileInfo(): void {
        try {
            if (typeof this.documentName === 'undefined') throw '';

            this.tileInfo = this.parseTileInfoFromDocName(this.documentName);
        } catch (error) {
            this.fail('Could not parse needed info from filename. Make sure it is named like "+33+126_17".');
        }
    }

    private parseTileInfoFromDocName(docName: string): ITileInfo {
        const regex = /([+-]\d{1,3})/g;
        const latLon: RegExpMatchArray | null = docName.match(regex);
        const zoom: string | undefined = docName.split('_')[1];

        if (latLon === null || latLon?.length <= 1 || typeof zoom === 'undefined') throw '';

        return {
            lat: parseInt(latLon[0], 10),
            lon: parseInt(latLon[1], 10),
            zoom: parseInt(zoom, 10),
        };
    }

    private groupTexByZoom(): void {
        const layers: Layers = app.activeDocument.layers;

        // Type 'Layers' is not an array type or a string type or does not have a '[Symbol.iterator]()' method that returns an iterator.
        // eslint-disable-next-line @typescript-eslint/prefer-for-of
        for (let i = 0; i < layers.length; i++) {
            const texInfo: ITexInfo | null = this.getLayerInfo(layers[i]);

            if (texInfo === null) continue;

            // eslint-disable-next-line no-prototype-builtins
            if (!this.zoomGroups.hasOwnProperty(texInfo.tex_zoom.toString())) this.zoomGroups[texInfo.tex_zoom] = [];

            this.zoomGroups[texInfo.tex_zoom].push(layers[i]);
        }
    }

    private processTextures(): void {
        this.setHistoryState(this.PROC_START_SNAPSHOT_NAME);

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        for (let i: number = this.tileInfo!.zoom; i <= this.MAX_ZOOM; i++) {
            // eslint-disable-next-line no-prototype-builtins
            if (!this.zoomGroups.hasOwnProperty(i.toString())) continue;

            const zoomGroup: Layer[] = this.zoomGroups[i];

            for (const layer of zoomGroup) {
                this.setActiveSnapshot(this.PROC_START_SNAPSHOT_NAME);
                this.clearCanvas(layer, i);

                app.activeDocument.crop(layer.bounds);

                if (i !== this.tileInfo?.zoom) this.resizeCanvas(i);

                this.exportCanvas(layer);
            }
        }
    }

    private exportCanvas(layer: Layer): void {
        const document: Document = app.activeDocument;
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const ortho4xpName: string = this.documentName!.split('_')[0];
        const providerName: string = this.getProviderDirectoryName(layer.name);
        const destination = new Folder(`${document.path.toString()}/orthomagic4ps/${ortho4xpName}/${providerName}`);

        if (!destination.exists) destination.create();

        const file: File = new File(`${destination.toString()}/${layer.name}.jpg`);

        document.saveAs(file, this.EXPORT_OPTIONS, true, Extension.LOWERCASE);
    }

    private getProviderDirectoryName(name: string): string {
        const nameParts: string[] = name.split('_');
        const providerPart: string = nameParts[2];
        const provider: string = providerPart.slice(0, -2);
        const zoom: string = providerPart.slice(-2);

        return `${provider}_${zoom}`;
    }

    /** It's faster to delete all other layers on every iteration and then crop and resize,
     * than cropping and resizing the document with all layers/smart objects for every iteration.
     * Also deleting all invisible layers at once is very much faster than deleting every layer on its own.
     */
    private clearCanvas(currentLayer: Layer, currentZoom: number): void {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        for (let i: number = this.tileInfo!.zoom; i <= this.MAX_ZOOM; i++) {
            // eslint-disable-next-line no-prototype-builtins
            if (!this.zoomGroups.hasOwnProperty(i.toString())) continue;

            const zoomGroup: Layer[] = this.zoomGroups[i];

            for (const layer of zoomGroup) {
                if (
                    layer.name === currentLayer.name ||
                    (i > currentZoom && this.isLayerOverlapping(currentLayer, layer))
                )
                    continue;

                // If there are linked layers, remove them as well.
                // Special case for using cut masks - see wiki.
                if (layer.linkedLayers.length > 0)
                    for (const linkedLayer of layer.linkedLayers) linkedLayer.visible = false;

                layer.visible = false;
            }
        }

        this.deleteHiddenLayers();
    }

    private isLayerOverlapping(currentLayer: Layer, overlapLayer: Layer): boolean {
        const cLeft: number | UnitValue = currentLayer.bounds[0];
        const cTop: number | UnitValue = currentLayer.bounds[1];
        const cRight: number | UnitValue = currentLayer.bounds[2];
        const cBottom: number | UnitValue = currentLayer.bounds[3];

        const oLeft: number | UnitValue = overlapLayer.bounds[0];
        const oTop: number | UnitValue = overlapLayer.bounds[1];
        const oRight: number | UnitValue = overlapLayer.bounds[2];
        const oBottom: number | UnitValue = overlapLayer.bounds[3];

        return cLeft < oRight && oLeft < cRight && cTop < oBottom && oTop < cBottom;
    }

    private deleteHiddenLayers(): void {
        const descriptor: ActionDescriptor = new ActionDescriptor();
        const reference: ActionReference = new ActionReference();

        reference.putEnumerated(this.sTID('layer'), this.sTID('ordinal'), this.sTID('hidden'));
        descriptor.putReference(this.sTID('null'), reference);

        app.executeAction(this.sTID('delete'), descriptor, DialogModes.NO);
    }

    /** Each layer needs to be resized separately after crop, otherwise the document would easily end up
     * with several hundred thousand pixels dimensions - and Photoshop does not like that
     */
    private resizeCanvas(zoom: number): void {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const zoomDiff: number = zoom - this.tileInfo!.zoom;
        const multiplier: number = Math.pow(2, zoomDiff);
        const width: number | UnitValue = app.activeDocument.width;
        const height: number | UnitValue = app.activeDocument.height;

        app.activeDocument.resizeImage(<number>width * multiplier, <number>height * multiplier);
    }

    private setHistoryState(name: string): void {
        const descriptor: ActionDescriptor = new ActionDescriptor();
        const reference1: ActionReference = new ActionReference();
        const reference2: ActionReference = new ActionReference();

        reference1.putClass(this.cTID('SnpS'));
        descriptor.putReference(this.cTID('null'), reference1);
        reference2.putProperty(this.cTID('HstS'), this.cTID('CrnH'));
        descriptor.putReference(this.cTID('From'), reference2);
        descriptor.putString(this.cTID('Nm  '), name);
        descriptor.putEnumerated(this.cTID('Usng'), this.cTID('HstS'), this.cTID('FllD'));

        app.executeAction(this.cTID('Mk  '), descriptor, DialogModes.NO);
    }

    private clearHistoryStates(): void {
        this.setActiveSnapshot(this.PROC_START_SNAPSHOT_NAME);

        // Type 'HistoryStates' is not an array type or a string type or does not have a '[Symbol.iterator]()' method that returns an iterator.
        // eslint-disable-next-line @typescript-eslint/prefer-for-of
        for (let i = 0; i < app.activeDocument.historyStates.length; i++) {
            const state: HistoryState = app.activeDocument.historyStates[i];

            if (!state.snapshot) continue;

            if (state.name === this.PROC_START_SNAPSHOT_NAME) this.removeSnapshot(state.name);
        }
    }

    private getSnapshot(name: string): HistoryState | undefined {
        // Ugh, Adobe.
        try {
            return app.activeDocument.historyStates.getByName(name);
        } catch (error) {
            return undefined;
        }
    }

    private setActiveSnapshot(name: string): void {
        const snapshot: HistoryState | undefined = this.getSnapshot(name);

        if (typeof snapshot !== 'undefined') app.activeDocument.activeHistoryState = snapshot;
    }

    private removeSnapshot(name: string): void {
        const descriptor: ActionDescriptor = new ActionDescriptor();
        const reference: ActionReference = new ActionReference();

        reference.putName(this.cTID('SnpS'), name);
        descriptor.putReference(this.cTID('null'), reference);

        app.executeAction(this.cTID('Dlt '), descriptor, DialogModes.NO);
    }

    private getLayerInfo(layer: Layer): ITexInfo | null {
        const splitName: string[] = layer.name.split('_');

        // Exactly three parts are expected
        if (splitName.length !== 3) return null;

        const tex_y: number = parseInt(splitName[0], 10);
        const tex_x: number = parseInt(splitName[1], 10);
        const tex_zoom: number = parseInt(splitName[2].slice(-2), 10);

        // All three parts are expected to be numbers
        if (isNaN(tex_x) || isNaN(tex_y) || isNaN(tex_zoom)) return null;

        // Some sanity check for supported zoom levels
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        if (tex_zoom < this.tileInfo!.zoom || tex_zoom > 19) return null;

        return {tex_x, tex_y, tex_zoom};
    }

    private fail(error: string): void {
        this.restoreRulerUnits();
        this.clearHistoryStates();

        throw new Error(error);
    }

    private finish(): void {
        this.restoreRulerUnits();
        this.clearHistoryStates();

        alert(`Operation finished.`);
    }
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const exportMagic = new ExportMagic();
